keyboards/rura66/rev1/oled_display.c

```diff
@@ -0,0 +1,113 @@
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+#include <stdio.h>
+
+// Defines names for use in layer keycodes and the keymap
+enum layer_names {
+  _QWERTY = 0,
+  _LOWER,
+  _RAISE,
+  _ADJUST
+};
+
+#ifdef OLED_DRIVER_ENABLE
+
+void render_status(void) {
+
+  // Render to mode icon
+  static const char os_logo[][2][3] PROGMEM  ={{{0x95,0x96,0},{0xb5,0xb6,0}},{{0x97,0x98,0},{0xb7,0xb8,0}}};
+  if (is_mac_mode()) {
+    oled_write_P(os_logo[0][0], false);
+    oled_write_P(PSTR("\n"), false);
+    oled_write_P(os_logo[0][1], false);
+  }else{
+    oled_write_P(os_logo[1][0], false);
+    oled_write_P(PSTR("\n"), false);
+    oled_write_P(os_logo[1][1], false);
+  }
+
+  oled_write_P(PSTR(" "), false);
+
+  // Host Keyboard Layer Status
+  oled_write_P(PSTR("Layer: "), false);
+
+  switch (get_highest_layer(layer_state)) {
+      case _QWERTY:
+          oled_write_P(PSTR("Default\n"), false);
+          break;
+      case _RAISE:
+          oled_write_P(PSTR("Raise\n"), false);
+          break;
+      case _LOWER:
+          oled_write_P(PSTR("Lower\n"), false);
+          break;
+      case _ADJUST:
+          oled_write_P(PSTR("Adjust\n"), false);
+          break;
+      default:
+          // Or use the write_ln shortcut over adding '\n' to the end of your string
+          oled_write_ln_P(PSTR("Undefined"), false);
+  }
+
+  oled_write_P(PSTR("\n"), false);
+
+  // Host Keyboard LED Status
+  led_t led_state = host_keyboard_led_state();
+  oled_write_P(led_state.num_lock ? PSTR("NUM ") : PSTR("    "), false);
+  oled_write_P(led_state.caps_lock ? PSTR("CAP ") : PSTR("    "), false);
+  oled_write_P(led_state.scroll_lock ? PSTR("SCR ") : PSTR("    "), false);
+}
+
+
+static void render_logo(void) {
+    static const char PROGMEM qmk_logo[] = {
+        0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8A, 0x8B, 0x8C, 0x8D, 0x8E, 0x8F, 0x90, 0x91, 0x92, 0x93, 0x94,
+        0xA0, 0xA1, 0xA2, 0xA3, 0xA4, 0xA5, 0xA6, 0xA7, 0xA8, 0xA9, 0xAA, 0xAB, 0xAC, 0xAD, 0xAE, 0xAF, 0xB0, 0xB1, 0xB2, 0xB3, 0xB4,
+        0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF, 0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0x00
+    };
+
+    oled_write_P(qmk_logo, false);
+}
+
+static void render_rgbled_status(bool full) {
+#ifdef RGBLIGHT_ENABLE
+  char buf[30];
+  if (RGBLIGHT_MODES > 1 && rgblight_is_enabled()) {
+      if (full) {
+          snprintf(buf, sizeof(buf), " LED %2d: %d,%d,%d ",
```

Comment:
Just a heads up, `sprintf` and `snprintf` are rather expensive, in termas of firmware size and MCU resources.  It can take up 1.5kb.  

While this is more verbose, it uses up a lot less firmware space:
```c
        uint16_t m = rgblight_get_mode();
        string[3] = '\0';
        string[2] = '0' + m % 10;
        string[1] = ( m /= 10) % 10 ? '0' + (m) % 10 : (m / 10) % 10 ? '0' : ' ';
        string[0] =  m / 10 ? '0' + m / 10 : ' ';
        oled_write_P(PSTR("Conf:"), false);
        oled_write(string, false);
        uint16_t h = rgblight_get_hue()/RGBLIGHT_HUE_STEP;
        string[3] = '\0';
        string[2] = '0' + h % 10;
        string[1] = ( h /= 10) % 10 ? '0' + (h) % 10 : (h / 10) % 10 ? '0' : ' ';
        string[0] =  h / 10 ? '0' + h / 10 : ' ';
        oled_write_P(PSTR(","), false);
        oled_write(string, false);
        uint16_t s = rgblight_get_sat()/RGBLIGHT_SAT_STEP;
        string[3] = '\0';
        string[2] = '0' + s % 10;
        string[1] = ( s /= 10) % 10 ? '0' + (s) % 10 : (s / 10) % 10 ? '0' : ' ';
        string[0] =  s / 10 ? '0' + s / 10 : ' ';
        oled_write_P(PSTR(","), false);
        oled_write(string, false);
        uint16_t v = rgblight_get_val()/RGBLIGHT_VAL_STEP;
        string[3] = '\0';
        string[2] = '0' + v % 10;
        string[1] = ( v /= 10) % 10 ? '0' + (v) % 10 : (v / 10) % 10 ? '0' : ' ';
        string[0] =  v / 10 ? '0' + v / 10 : ' ';
        oled_write_P(PSTR(","), false);
        oled_write(string, false);
        oled_write_ln_P(PSTR("\n     MOD HUE SAT VAR"), false);
```

Reply:
I did about 2 hours of research into this googling parts of what you wrote but couldn't find an example of this being used. Unsure how to get this to work or compile. I gave it my best shot but nothing. I also attempted to find other keyboards that use these strings but was unsuccessful in finding something to match or how to bring it into this keyboard. Any help is appreciated.

Reply:
There are a number of keymaps that are using this sort of notation in the repo. Not for the rgb values specifically,   I use it personally.

Reply:
Oh, and `get_u8_str(int, ' ')` will convert int to a 3 character string.